firebase-app-distribution/src/main/java/com/google/firebase/appdistribution/FirebaseAppDistribution.java

```diff
@@ -119,7 +134,107 @@ public void onFailure(@NonNull @NotNull Exception e) {
+                            FirebaseAppDistributionTesterApiClient client =
+                                new FirebaseAppDistributionTesterApiClient(fid, appId, apiKey);
+
+                            handleLatestReleaseFromClient(client, authToken);
+                          }
+                        });
+              }
+            })
+        .addOnFailureListener(
+            new OnFailureListener() {
+              @Override
+              public void onFailure(@NonNull Exception e) {
+                setCheckForUpdateTaskCompletionError(
+                    new FirebaseAppDistributionException(
+                        FirebaseAppDistributionException.Status.UNKNOWN));
+              }
+            });
+
+    return checkForUpdateTaskCompletionSource.getTask();
+  }
+
+  private void handleLatestReleaseFromClient(
+      FirebaseAppDistributionTesterApiClient client, String authToken) {
+    checkForUpdateExecutor.execute(
+        new Runnable() {
+          @Override
+          public void run() {
+            try {
+              AppDistributionRelease latestRelease = client.fetchLatestRelease(authToken);
+
+              Context context = firebaseApp.getApplicationContext();
+              long currentInstalledVersionCode = getInstalledAppVersionCode(context);
+
+              //move logic outside,
+              checkForUpdateHandler.post(
+                  new Runnable() {
+                    @Override
+                    public void run() {
+                      if (currentInstalledVersionCode
+                          < Long.parseLong(latestRelease.getBuildVersion())) {
+                        // latest release is newer
+                        checkForUpdateTaskCompletionSource.setResult(latestRelease);
+                      } else {
+                        checkForUpdateTaskCompletionSource.setResult(null);
+                      }
+                    }
+                  });
+            } catch (Exception e) {
+              setCheckForUpdateTaskCompletionError(
+                  new FirebaseAppDistributionException(
+                      FirebaseAppDistributionException.Status.UNKNOWN));
+            }
+          }
+        });
+  }
+
+  private long getInstalledAppVersionCode(Context context) {
+    PackageInfo pInfo = null;
+    try {
+      pInfo = context.getPackageManager().getPackageInfo(context.getPackageName(), 0);
+    } catch (PackageManager.NameNotFoundException e) {
+      checkForUpdateTaskCompletionSource.setException(
+          new FirebaseAppDistributionException(FirebaseAppDistributionException.Status.UNKNOWN));
+    }
+    return PackageInfoCompat.getLongVersionCode(pInfo);
+  }
+
+  private void setCheckForUpdateTaskCompletionError(FirebaseAppDistributionException e) {
+    if (checkForUpdateTaskCompletionSource != null
+        && !checkForUpdateTaskCompletionSource.getTask().isComplete()) {
+      this.checkForUpdateTaskCompletionSource.setException(e);
+    }
```

Comment:
In general, it's good to have some logical ordering to methods in Java classes. There's no prescriptive way to order methods, but one that I've seen is 1. all public methods 2. All package-private or private methods in the order they are first called https://google.github.io/styleguide/javaguide.html#s3.4.2-ordering-class-contents